CORPORATION: Corp changes prior to 2.3 finalization

Snarling opened the pull request:
Fixes #491
Rewrote various corp code.

Changed property names and in some cases changed t he internal data structures so they make more sense. The corp changes in 2.3 require a full reset of the player's corporation due to the number of internal changes, so this provides the ideal time to improve the property names.

**There are some Corporation API changes, mostly due to property name changes:**

Usage changes:
* setProductMarketTA1 and setProductMarketTA2 no longer take in a city name, as this was not being used.

Changes due to one-time unlockable upgrades being renamed from "unlock upgrades" to "unlocks":
* hasUnlockUpgrade -> hasUnlock
* getUnlockUpgradeCost -> getUnlockCost
* unlockUpgrade -> purchaseUnlock

Properties on Division object:
* prodMult -> productionMult
* research -> researchPoints
* upgrades: Removed
* numAdVerts: Added (this was previously the only info provided on "upgrades")

Properties on Export object:
* amt -> amount
* loc -> city
* div -> division

Properties on Material object:
* cmp -> competition
* cost -> marketPrice
* dmd -> demand
* exp -> exports
* prod -> productionAmount
* qlt -> quality
* qty -> stored
* sAmt -> desiredSellAmount
* sCost -> desiredSellPrice
* sell -> actualSellAmount

Properties on Office object:
* avgEne -> avgEnergy
* avgMor -> avgMorale
* employeeProd -> employeeProductionByJob
* employees -> numEmployees
* loc -> city
* maxEne -> maxEnergy
* maxMor -> maxMorale

Properties on Product object:
* cmp -> competition
* dmd -> demand
* effRat -> effectiveRating
* pCost -> productionCost
* prod -> productionAmount
* properties -> stats
  * qlt -> quality
  * per -> performance
  * dur -> durability
  * rel -> reliability
  * aes -> aesthetics
  * fea -> features
* qty -> stored
* rat -> rating
* sAmt -> desiredSellAmount
* sCost -> desiredSellCost
* sell -> actualSellAmount

Properties on Warehouse object:
* loc -> city

Other misc changes:
* Removed JoyWire and Bulk Purchasing from research names list, these were removed at some point in the 2.3 corp overhaul.

Snarling committed:
Clear divisions on migration + divisions as JSONMap

d0sboots commented on src/Achievements/Achievements.ts:
```diff
@@ -447,17 +447,25 @@ export const achievements: Record<string, Achievement> = {
         let totalEmployees = 0;
-        for (const o of Object.values(d.offices)) if (o && o.totalEmployees) totalEmployees += o.totalEmployees;
+        for (const office of Object.values(division.offices)) {
+          if (office && office.totalEmployees) totalEmployees += office.totalEmployees;
```
Optional: For stuff like this, you can also do `totalEmployees += office?.totalEmployees ?? 0`

d0sboots commented on src/Corporation/ui/modals/ExportModal.tsx:
```diff
@@ -102,18 +101,18 @@ export function ExportModal(props: IProps): React.ReactElement {
-              {division.name}
+      <Select onChange={onIndustryChange} value={division.name}>
+        {[...corp.divisions.values()]
+          .filter((div) => isRelevantMaterial(props.mat.name, div))
```
I'd call this something else. "div" is overloaded with HTML div, especially in a JSX context.

Snarling replied:
Yeah, related to the other comment about "division" being used as a frequent variable name.

d0sboots commented on src/Corporation/ui/modals/ExportModal.tsx:
```diff
@@ -25,24 +25,26 @@ interface IProps {
   if (Object.keys(defaultDivision.warehouses).length === 0)
     throw new Error("Export popup created in a division with no warehouses.");
-  const [industry, setIndustry] = useState<string>(defaultDivision.name);
+  const [division, setDivision] = useState(defaultDivision);
```
Maybe call this currentDivision? There's too many other places where "division" is wanted.

d0sboots commented on src/Corporation/ui/modals/SellDivisionModal.tsx:
```diff
@@ -21,39 +19,40 @@ interface IProps {
-  const allIndustries = Object.values(corp.divisions).sort();
-  const [industry, setIndustry] = useState(allIndustries[0]);
+  const allDivisions = [...corp.divisions.values()];
+  const [division, setDivision] = useState(allDivisions[0]);
```
Same thing about naming

d0sboots reviewed (approved):
Looks good!

Snarling commented:
Ughhhhhhh I had a bunch more in progress changes but had not committed them due to not resolving all the build errors yet (due to data structure changes). And my drive that I normally do development work on is not working. I just set up a new dev environment on my other drive, so I lost these staged changes.

The new way to migrate a corp will be check if corp exists and divisions is an array (or some other reliable indication that it's a loaded oldcorp) -> if so, create a generous valuation of the corp and just replace the entire corp with a new one with that amount of money. This allows any and all structural changes in corp without requiring careful management of save migration. I'll reimplement some of the more important structural changes that I had in the in progress changes that got lost, then take care of the React stuff I never got around to.

d0sboots commented:
> The new way to migrate a corp will be check if corp exists and divisions is an array (or some other reliable indication that it's a loaded oldcorp) -> if so, create a generous valuation of the corp and just replace the entire corp with a new one with that amount of money.

I think that made it into this version of the PR?

Snarling commented:
> I think that made it into this version of the PR?

The current PR version only replaces the `divisions` property and keeps the rest of the corp object intact, I plan on changing it so that the corp object is not kept at all (and is just assessed for a value before replacing it with a completely fresh corp object)

d0sboots commented:
We'll want a big warning somewhere saying that you should put off updating if you have an in-progress corp; losing all your research means you lose most of your progress that actually takes time.

Snarling committed:
Bunch of corp changes

Still in progress on making changes. This was just a stopping point where no build errors are present.

Snarling commented:
Still working on this... I underestimated how much of a mess corp code was, and I haven't had as much time as I'd like this week.

Committed just because I had a stopping point with no build errors, will work some more on this tonight.

Snarling committed:
Lint, format, remove more corp load code

d0sboots commented:
Ahhhh, the new names are so much better! No more stupid `sllman`!

Snarling committed:
Last savegame structure changes?

Snarling commented:
Alright I think I'm done with changes that will affect savegame format.

Of note, something that seems like a weirdness but is intentional is that in some cases I am using JSONMap, in others I am using records. The JSONMaps are being used in instances where the keys are arbitrary strings (player-decided names), where a map avoids any weirdness with own keys vs prototype keys. None of these objects are of a scale where there should be much performance difference either way.

Just a couple react issues to clear up now.

Snarling committed:
Fix corp react errors
